### DISCUSSION TITLE
"Optimum = 0" when printing solution output

### DISCUSSION THREAD

**Discussion by lthUniBonn** [mdolab/pyoptsparse#241] [Q&A] [answered]
Hello everyone, 

I'm starting my work with optimizers with pyOptSparse at the moment. Therefore I tried to understand to complete output given from the optimizer, where I stumbled upon the "Optimum" output for the objective. I for example get:
```commandline
   Objectives
      Index  Name            Value          Optimum
          0  obj    -9.692803E-02     0.000000E+00
```
This optimum seems to be initialized to 0 and then never changed in the pyOptSparse repo code - as I'v checked the pyOptSparse code with `grep -ri optimum *` and get the output:
```commandline
doc/quickstart.rst:         Index  Name            Value          Optimum
doc/quickstart.rst:         Index  Name            Value          Optimum
pyoptsparse/pyOpt_optimization.py:        text += fmt.format("Index", "Name", "Value", "Optimum", width=num_c)
pyoptsparse/pyOpt_optimization.py:            text += fmt.format(idx, obj.name, value, obj.optimum, width=num_c)
pyoptsparse/pyNOMAD/pyNOMAD.py:# Description: This program finds the optimum solution for a given objective
**pyoptsparse/pyOpt_objective.py:        self.optimum = 0.0**
pyoptsparse/pyOpt_objective.py:        res = "        Name        Value        Optimum\n"
pyoptsparse/pyOpt_objective.py:        res += "%12g  %12g\n" % (np.real(self.value), np.real(self.optimum))
pyoptsparse/pyPSQP/source/pqsubs.f:*         MFP=2-OPTIMUM FEASIBLE POINT. MFP=3-REPEATED SOLUTION.
pyoptsparse/pyPSQP/source/pqsubs.f:*         ITERL=2-OPTIMUM FEASIBLE POINT. ITERL=-1 FEASIBLE POINT DOES
pyoptsparse/pyPSQP/source/pqsubs.f:*         NOT EXISTS. ITERL=-2 OPTIMUM FEASIBLE POINT DOES NOT EXISTS.
pyoptsparse/pyPSQP/source/pqsubs.f:*     OPTIMUM ON A LINEAR MANIFOLD OBTAINED
pyoptsparse/pyPSQP/source/pqsubs.f:*         MFP=2-OPTIMUM FEASIBLE POINT. MFP=3-REPEATED SOLUTION.
pyoptsparse/pyPSQP/source/pqsubs.f:*         ITERQ=2-OPTIMUM FEASIBLE POINT. ITERQ=-1 FEASIBLE POINT DOES
pyoptsparse/pyPSQP/source/pqsubs.f:*         NOT EXISTS. ITERQ=-2 OPTIMUM FEASIBLE POINT DOES NOT EXISTS.
test/test_rosenbrock.py:    #  At the optimum, the function is f(x) = 0.
test/test_rosenbrock.py:            # check optimum stored in history file against xstar
test/test_hs071.py:        # since we restarted from the optimum
test/test_hs015.py:        # we check either optimum via try/except
test/test_hs015.py:            # check optimum stored in history file against xstar
```

An Optimum of 0 seems odd for a minimizer, that can also go way down the negative function values for the objective. But it also does not seem to impact any part of the optimization, as I don't find it in the code...
So can you maybe give more insight into the meaning here? 
(Just for information: I'm using the SLSQP optimizer you provide)

Thanks a lot for providing this tool :)
Cheers,
Lavinia

**Comment by marcomangano** [comment, marked as answer]
Hello Lavinia,
It looks like you actually caught a small bug that got undetected for some time! That additional `Optimum` printout is misleading and it is not providing any useful information, we will remove it soon (see [this PR](https://github.com/mdolab/pyoptsparse/pull/242)).

The objective function value at the final iteration is provided in the `Value` column. Depending on the optimizer, the additional printout files will include more information on the optimization path and parameters. You can also use [optview](https://mdolab-pyoptsparse.readthedocs-hosted.com/en/latest/postprocessing.html) to visualize the data from the history file (whose name is defined via the `storeHistory` attribute in the `Optimizer` instance, eg in the example scripts: `sol = opt(optProb, sens=.., storeHistory="<yourHistoryfilename>.hst")`).

Hope this helps.
Cheers, 

Marco

**Reply by lthUniBonn** [reply]
Hi Marco, 

yes, this makes it all much clearer now - and you already implemented it, that was fast! 
I wanted to check out optview as well, when my optimization was set up - so I just need to add the storeHistory argument, that's nice and easy :) 
So thanks for the quick help! 
Cheers, 
Lavinia

**Reply by marcomangano** [reply]
No problem, glad it helped.
We will add that tip on the history file to the docs soon, it should have been there already.
Let us know if you have any suggestions about how to improve the documentation as you try more features.

Good luck!
Marco

**Reply by lthUniBonn** [reply]
Thanks, will do! 

Cheers, 
Lavinia
